zebra-state/src/in_memory/chains.rs

```diff
@@ -0,0 +1,37 @@
+use super::Error;
+use crate::{Request, Response};
+use futures::prelude::*;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower::Service;
+
+/// A service wrapper that tracks multiple chains, handles reorgs, and persists
+/// blocks to disk once they're past the reorg limit
+pub(crate) struct ChainsState<S> {
+    /// The inner state service that only tracks a single chain
+    inner: S,
+}
+
+impl<S> Service<Request> for ChainsState<S> {
+    type Response = Response;
+    type Error = Error;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, req: Request) -> Self::Future {
+        tracing::debug!(?req);
+        match req {
+            Request::AddBlock { block } => todo!(),
+            Request::GetBlock { hash } => todo!(),
+            Request::GetTip => todo!(),
+            Request::GetDepth { hash } => todo!(),
+            Request::GetBlockLocator { genesis } => todo!(),
```

Comment:
If we want the state to support reorgs, we'll need slightly different variants here:
```suggestion
            Request::SetMainTip { hash } => todo!(),
            Request::GetMainTip => todo!(),
            Request::GetTips => todo!(),
            Request::GetMainDepth { hash } => todo!(),
            Request::GetMainBlockLocator { genesis } => todo!(),
```

Reply:
Do you think the state should add and update tips during AddBlock?
Or should we add extra variants to manage side-chain tips?

I think it will be cleaner to do it at the state level, because updates can use the current state directly. (Which is much harder at the zebra-consensus level, because we'd need to use some kind of transaction.)

Reply:
I was thinking this would be entirely done internally, so `SetMainTip` would be a side effect of `AddBlock`.

Reply:
To find the main tip, we need to calculate the cumulative proof of work of each chain. And we need to break ties using block download time (or something similar, like verification time).

I feel like we're splitting a lot of verification logic between zebra-consensus and zebra-state, and that coupling will make it hard to have good interfaces.

For example, the individual block proof of work is also used by zebra-consensus to check that the block passes the difficulty threshold.

Reply:
I agree, this is me mixing up the responsibilities again, I always forget what is consensus critical and what isn't. I still don't love how intertwined these are and how keeping the responsibilities separate means this logic has to get split up into multiple stages but it seems like the right call.

I think ideally it would be best if `zebra-consensus` could export a function to calculate the cumulative work and or pick a chain, but then have `zebra-state` call this function as part of its `AddBlock` routine to keep the API minimal. But I don't know how we'd be able to accomplish this without breaking the `zebra-consensus` crate into two crates where one is depended on by `zebra-state` and then the other depends on `zebra-state`, which doesn't seem any better than just having the more complex API live in `zebra-state`.

Damn you modularity.

Reply:
Also @teor2345 I think I'd like to move any extra `Request / Response` pairs that we add to the `chain-reorg-impl` PR, rather than this one. This was more about the broad strokes and getting approval on the design where I wrap the persistent single-chain state service with the volatile multi chain reorging state service

Reply:
I think we have three options for picking the tip:
* send the cumulative proof of work value and a timestamp tie-breaker value in the AddBlock request
* send a closure that returns a tip order as part of the AddBlock request
* add a SetTip request which is called by the verifier

Now that I think about it, the state should set the tip, because it is the only place with access to all the up-to-date data. (The verifier can query the state, but the state can change between GetBlock and SetTip. Which is a nasty race condition.)

I don't think it's worth making AddBlock have generic data or a generic closure, because we're pretty sure we'll only ever use the cumulative work and a timestamp.

So here's one design:
* zebra-chain defines the ProofOfWork type and its Ord implementation
* zebra-state defines a ChainTipPriority type containing ProofOfWork and a timestamp, and its Ord implementation
* zebra-consensus calculates the ChainTipPriority value and sends it as part of AddBlock
* as a final tie-breaker, just in case we get two identical proof of work and timestamp values, zebra-state compares the block hashes

Reply:
I think it's ok for the state to know about hash, height, proof of work, and timestamp.

But we should avoid accessing any other fields in the state. (We could make the rest of the chain context generic or private.)

Reply:
> send a closure that returns a tip order as part of the AddBlock request

I think this is the solution that most appeals to me. We'd basically be making the `Ord` impl for our chains be caller provided.

> I don't think it's worth making AddBlock have generic data or a generic closure, 

I think we can do this while avoiding adding any generic parameters assuming that the closure wouldn't need to capture anything, we can just use function pointers, if `zebra-consensus` would need to capture state to send in the closure to correctly order the tips then we might have to box closures which starts to be a less than ideal API.

Reply:
> zebra-consensus calculates the ChainTipPriority value and sends it as part of AddBlock

Will `zebra-consensus` have access to the parent block already when its adding the block to the state layer? It seems like it would, this also lets me rip out all the checks for matching parent hashes, which in retrospect is pretty clearly the point. yay!

> zebra-state defines a ChainTipPriority type containing ProofOfWork and a timestamp, and its Ord implementation

This is ultimately a consensus rule, right? Seems less than ideal that the ord impl would have to live in zebra-state. I think im gonna try running with your second solution with passed in closures and see what kinda API I can come up with, everything else will be the same except `zebra-state` wouldn't define the ord impl, and it would be passed in from `zebra-consensus`.

Reply:
Capturing state in the closure would open up the possibility of using outdated information in the closure. Which could lead to similar race conditions to doing the comparison in zebra-consensus.

So we should make state capture impossible, by choosing the right closure type.

I think the closure can just take a copy of the chain context.

Reply:
you can't make captures impossible via any of the `Fn` traits, only by using fn pointers, so it sounds like function pointers are the way to go.

Reply:
I also like the idea of making the chain context generic, so the exact data it contains is managed by zebra-consensus. (And not visible to zebra-state.)

I'm thinking that each verifier in zebra-consensus might like to add its own context. For example, the block header will need time and difficulty, and transparent and shielded transactions will need their own particular UTXO or nullifier context.

Then we can roll all the different contexts up into a generic struct.

So the AddBlock request would include the block itself, and the chain context data for that block.

One nice property of this design is that the context is updated where it is used, rather than in a separate module. (And we can use im in zebra-consensus to save RAM.)

Another nice property is that the zebra-state design is a lot easier to understand, because we hide the context and its implementation. So the state just deals with hashes and heights and closures and tips and chains.

Reply:
> Will zebra-consensus have access to the parent block already when its adding the block to the state layer? It seems like it would, this also lets me rip out all the checks for matching parent hashes, which in retrospect is pretty clearly the point. yay!

Yes, we can't verify a block without its parent context (in BlockVerifier) or parent block (in CheckpointVerifier).

Reply:
> I also like the idea of making the chain context generic, so the exact data it contains is managed by zebra-consensus. (And not visible to zebra-state.)

We'll need this generic parameter as a requirement, because we'll need a different implementation for CheckpointVerifier and BlockVerifier:
* CheckpointVerifier: no context, compare height
* BlockVerifier: full context, compare cumulative work, timestamp, hash

I think we will also want different contexts for different network updates in BlockVerifier.

Reply:
> I also like the idea of making the chain context generic, so the exact data it contains is managed by zebra-consensus. (And not visible to zebra-state.)

OO, how about a `trait ChainState: Ord + ... { ... }` defined in `zebra-state`, that way everything can live in consensus and we can only require the interface to info we care about like `fn block(&self) -> Arc<Block>`, etc. That also avoids needing to put a closure in the `Request` sent to the state service, which I think improves the API even more.

> One nice property of this design is that the context is updated where it is used, rather than in a separate module. (And we can use im in zebra-consensus to save RAM.)

Not sure I'm following this point, particularly the `im` in `zebra-consensus` usage. My expectation was that the `ChainState` type would still be a per height type, and the chains themselves would still live in `zebra-state`, and the chains themselves are the only place I'd think `im` would apply.

Reply:
I think the ChainState should be generic, because then we can use the type system to manage transitions between the checkpoint chain state, and the chain state for each network upgrade.

Reply:
> > One nice property of this design is that the context is updated where it is used, rather than in a separate module. (And we can use im in zebra-consensus to save RAM.)
> 
> Not sure I'm following this point, particularly the im in zebra-consensus usage. My expectation was that the ChainState type would still be a per height type, and the chains themselves would still live in zebra-state, and the chains themselves are the only place I'd think im would apply.

Here's how I think we'll use im inside ChainState itself:
- block.header.time: an im vector of recent block times
- block.header.bits: an im vector of recent block difficulties
- light_client_hash: an im vector of all subtree roots (Heartwood only)
- other state required to validate transactions (as im data structures)
https://github.com/ZcashFoundation/zebra/pull/743/files#r459920766

I think we might also use im sets for the UTXO, shielded output, and nullifier deltas.

Reply:
For the block header time and bits those are already available in the block itself right? I think the missing piece of info for me is an understanding of exactly how much context from the chain the verifier will need. Like, I don't understand why we'd need to access multiple recent block times, and for the block difficulties the way I planned on implementing it was just storing the cumulative work of the chain leading up to the current chain state in the chain state, and then you do that + chainstate.block.header.bits to create the next chain state's cummulative work.

Reply:
You need recent block times to compute difficulty adjustments (https://zips.z.cash/protocol/protocol.pdf#diffadjustment).

Reply:
I'd like the code to be really explicit about the data used for verification. And I'd like all that data to be available in response to a single context query. (As much as possible - we might need to check deltas, then query larger sets of data on disk.)

Some information is only relevant to a particular network upgrade. For each network upgrade, I'd like to use a context struct that only contains the relevant information. So I'm hoping that we can make the context type generic.

If we use a context type for each network upgrade, then transitions between network upgrades will be enforced by the type system. (There can also be a different code path for the transition block itself).

Here's some of what we'll need:

>     * block.header.time: an im vector of recent block times

We need recent times to compute the valid time range for child blocks.

>     * block.header.bits: an im vector of recent block difficulties

We need recent difficulties to compute the minimum difficulty threshold for child blocks. And as @daira said, we need both times and difficulties to compute difficulty adjustments.

We need this info for `PoWAveragingWindow = 17` or `PoWMedianBlockSpan = 11` recent blocks.

>     * light_client_hash: an im vector of all subtree roots (Heartwood only)

When we are validating Heartwood blocks, we need a list of Merkle Mountain Range subtree roots:
https://zips.z.cash/zip-0221#background

The older roots will need to be stored separately as part of the disk state, so we don't have to scan the whole Heartwood chain on startup get the full set of roots. The set of roots is small (less than `log(N)`, where N is the number of Heartwood blocks).

Most of the changes to the roots will happen in recent blocks, so we can use an `im::Vector` here.

Note that this data is not required for pre-Heartwood blocks.

> I think we might also use im sets for the UTXO, shielded output, and nullifier deltas.

I think we'll be using @yaahc's `im` delta maps from `hash => Added | Removed` for all these sets, because they are large.

Reply:
I've changed my mind, I don't think we should duplicate the times or bits in both the recent blocks, and the context data structure.

I still think we'll need the heartwood light client root set.

Reply:
Let's continue here?
https://github.com/ZcashFoundation/zebra/pull/743#discussion_r460356214